<launch>

    <!--  ************** Global Parameters ***************  -->
    <param name="/use_sim_time" value="true"/>
    <arg name="controller"  default="mpc" doc="opt: dwa, mpc, pure_pursuit"/> 
    <arg name="map"  default="empty" doc="opt: empty"/> 

    <arg name="r1_x_pos" default="3.5"/>
    <arg name="r1_y_pos" default="3.5"/>
    <arg name="r1_z_pos" default="0.0"/>
    <arg name="r1_roll" default="0.0"/>
    <arg name="r1_pitch" default="0.0"/>
    <arg name="r1_yaw" default="3.92"/>

    <arg name="r2_x_pos" default="-3.5"/>
    <arg name="r2_y_pos" default="3.5"/>
    <arg name="r2_z_pos" default="0.0"/>
    <arg name="r2_roll" default="0.0"/>
    <arg name="r2_pitch" default="0.0"/>
    <arg name="r2_yaw" default="-0.893008"/>

    <arg name="r3_x_pos" default="3.5"/>
    <arg name="r3_y_pos" default="-3.5"/>
    <arg name="r3_z_pos" default="0.0"/>
    <arg name="r3_roll" default="0.0"/>
    <arg name="r3_pitch" default="0.0"/>
    <arg name="r3_yaw" default="2.35619"/>


    <arg name="r4_x_pos" default="-4.0"/>
    <arg name="r4_y_pos" default="-4.0"/>
    <arg name="r4_z_pos" default="0.0"/>
    <arg name="r4_roll" default="0.0"/>
    <arg name="r4_pitch" default="0.0"/>
    <arg name="r4_yaw" default="0.749466"/>

    <arg name="robot1"  default="robot_0"/>
    <arg name="robot2" default="robot_1"/>
    <arg name="robot3"  default="robot_2"/>
    <arg name="robot4" default="robot_3"/>

    <!--  ************** Map Server **************  -->

    <node name="map_server" pkg="map_server" type="map_server" args="$(find gazebo_rl_test)/map/ractangle_10_map.yaml" output="screen" if="$(eval map == 'empty')">
        <param name="frame_id" value="/map"/>
    </node>


    <group ns = "$(arg robot1)">
        <param name="tf_prefix" value="$(arg robot1)" />


        <!--  ************** Localization **************  -->
        
        <node pkg="amcl" type="amcl" name="amcl" output="screen">
            <remap from="map" to="/map"/>

            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/3amcl_params.yaml" command="load" />
            <param name="initial_pose_x"            value="$(arg r1_x_pos)"/>
            <param name="initial_pose_y"            value="$(arg r1_y_pos)"/>
            <param name="initial_pose_a"            value="$(arg r1_yaw)" if="$(eval map == 'empty')"/>
            <remap from="/scan" to="$(arg robot1)/scan"/>

            
            <param name="global_frame_id" value="/map"/>
            <param name="odom_frame_id" value="$(arg robot1)/odom"/>
            <param name="base_frame_id" value="$(arg robot1)/base_footprint"/>    
        
        </node>


        <!--  ************** Navigation ***************  -->   
        <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
            <remap from="map" to="/map"/>

            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="global_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="local_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/local_costmap_params.yaml" command="load" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/global_costmap_params.yaml" command="load" />        
            
        <!-- Global Planner -->
            <param name="base_global_planner" value="global_planner/GlobalPlanner" />
            <param name="planner_frequency" value="1.0" />
            <param name="planner_patience" value="5.0" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/global_planner_params.yaml" command="load" />

            <param name="global_costmap/robot_base_frame" value="/$(arg robot1)/base_footprint"/>
            <param name="local_costmap/robot_base_frame" value="/$(arg robot1)/base_footprint"/>

            <!-- Local Planner -->
            <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS" />

            <!-- Our carlike robot is not able to rotate in place -->
            <param name="clearing_rotation_allowed" value="false" /> 

            <!-- external controller -->
            <remap from="cmd_vel" to="/fake_cmd" />    
        </node>
        
        <!--  ************** look a head **************  -->
        <node name="Look_a_head_Node" pkg="gazebo_rl_test" type="Look_a_head_Node" output="screen">
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/lookahead.yaml" command="load" />
            <param name="odom_topic"            value="/$(arg robot1)/odom"/>
            <param name="plan_topic"            value="/$(arg robot1)/move_base/GlobalPlanner/plan"/>
            <param name="goal_topic"            value="/$(arg robot1)/move_base_simple/goal"/>
            <param name="amcl_topic"            value="/$(arg robot1)/amcl_pose"/>
            <param name="lah_point_topic"       value="/$(arg robot1)/lookAhead_point"/>
            <param name="lah_marker_topic"      value="/$(arg robot1)/LookAhead/path_marker"/>
            <param name="odom_frame"            value="/$(arg robot1)/odom"/>

        </node>

    </group>

    <group ns = "$(arg robot2)">
        <param name="tf_prefix" value="$(arg robot2)" />


        <!--  ************** Localization **************  -->
        
        <node pkg="amcl" type="amcl" name="amcl" output="screen">
            <remap from="map" to="/map"/>

            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/3amcl_params.yaml" command="load" />
            <param name="initial_pose_x"            value="$(arg r2_x_pos)"/>
            <param name="initial_pose_y"            value="$(arg r2_y_pos)"/>
            <param name="initial_pose_a"            value="$(arg r2_yaw)" if="$(eval map == 'empty')"/>
            <remap from="/scan" to="$(arg robot2)/scan"/>
                
            <param name="global_frame_id" value="/map"/>
            <param name="odom_frame_id" value="$(arg robot2)/odom"/>
            <param name="base_frame_id" value="$(arg robot2)/base_footprint"/>    
        
        </node>

        <!--  ************** Navigation ***************  -->   
        <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
            <remap from="map" to="/map"/>
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="global_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="local_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/local_costmap_params.yaml" command="load" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/global_costmap_params.yaml" command="load" />        
            
        <!-- Global Planner -->
            <param name="base_global_planner" value="global_planner/GlobalPlanner" />
            <param name="planner_frequency" value="1.0" />
            <param name="planner_patience" value="5.0" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/global_planner_params.yaml" command="load" />

            <param name="global_costmap/robot_base_frame" value="/$(arg robot2)/base_footprint"/>
            <param name="local_costmap/robot_base_frame" value="/$(arg robot2)/base_footprint"/>

            <!-- Local Planner -->
            <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS" />

            <!-- Our carlike robot is not able to rotate in place -->
            <param name="clearing_rotation_allowed" value="false" /> 

            <!-- external controller -->
            <remap from="cmd_vel" to="/fake_cmd" />    
        </node>
        
        <!--  ************** look a head **************  -->
        <node name="Look_a_head_Node" pkg="gazebo_rl_test" type="Look_a_head_Node" output="screen">
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/lookahead.yaml" command="load" />
            <param name="odom_topic"            value="/$(arg robot2)/odom"/>
            <param name="plan_topic"            value="/$(arg robot2)/move_base/GlobalPlanner/plan"/>
            <param name="goal_topic"            value="/$(arg robot2)/move_base_simple/goal"/>
            <param name="amcl_topic"            value="/$(arg robot2)/amcl_pose"/>
            <param name="lah_point_topic"       value="/$(arg robot2)/lookAhead_point"/>
            <param name="lah_marker_topic"      value="/$(arg robot2)/LookAhead/path_marker"/>
            <param name="odom_frame"            value="/$(arg robot2)/odom"/>

        </node>

    </group>

    <group ns = "$(arg robot3)">
        <param name="tf_prefix" value="$(arg robot3)" />


        <!--  ************** Localization **************  -->
        
        <node pkg="amcl" type="amcl" name="amcl" output="screen">
            <remap from="map" to="/map"/>

            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/3amcl_params.yaml" command="load" />
            <param name="initial_pose_x"            value="$(arg r3_x_pos)"/>
            <param name="initial_pose_y"            value="$(arg r3_y_pos)"/>
            <param name="initial_pose_a"            value="$(arg r3_yaw)" if="$(eval map == 'empty')"/>
            <remap from="/scan" to="$(arg robot3)/scan"/>
                
            <param name="global_frame_id" value="/map"/>
            <param name="odom_frame_id" value="$(arg robot3)/odom"/>
            <param name="base_frame_id" value="$(arg robot3)/base_footprint"/>    
        
        </node>

        <!--  ************** Navigation ***************  -->   
        <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
            <remap from="map" to="/map"/>
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="global_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="local_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/local_costmap_params.yaml" command="load" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/global_costmap_params.yaml" command="load" />        
            
        <!-- Global Planner -->
            <param name="base_global_planner" value="global_planner/GlobalPlanner" />
            <param name="planner_frequency" value="1.0" />
            <param name="planner_patience" value="5.0" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/global_planner_params.yaml" command="load" />

            <param name="global_costmap/robot_base_frame" value="/$(arg robot3)/base_footprint"/>
            <param name="local_costmap/robot_base_frame" value="/$(arg robot3)/base_footprint"/>

            <!-- dLocal Planner -->
            <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS" />

            <!-- Our carlike robot is not able to rotate in place -->
            <param name="clearing_rotation_allowed" value="false" /> 

            <!-- external controller -->
            <remap from="cmd_vel" to="/fake_cmd" />    
        </node>
        
        <!--  ************** look a head **************  -->
        <node name="Look_a_head_Node" pkg="gazebo_rl_test" type="Look_a_head_Node" output="screen">
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/lookahead.yaml" command="load" />
            <param name="odom_topic"            value="/$(arg robot3)/odom"/>
            <param name="plan_topic"            value="/$(arg robot3)/move_base/GlobalPlanner/plan"/>
            <param name="goal_topic"            value="/$(arg robot3)/move_base_simple/goal"/>
            <param name="amcl_topic"            value="/$(arg robot3)/amcl_pose"/>
            <param name="lah_point_topic"       value="/$(arg robot3)/lookAhead_point"/>
            <param name="lah_marker_topic"      value="/$(arg robot3)/LookAhead/path_marker"/>
            <param name="odom_frame"            value="/$(arg robot3)/odom"/>

        </node>

    </group>

    <group ns = "$(arg robot4)">
        <param name="tf_prefix" value="$(arg robot4)" />


        <!--  ************** Localization **************  -->
        
        <node pkg="amcl" type="amcl" name="amcl" output="screen">
            <remap from="map" to="/map"/>

            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/3amcl_params.yaml" command="load" />
            <param name="initial_pose_x"            value="$(arg r4_x_pos)"/>
            <param name="initial_pose_y"            value="$(arg r4_y_pos)"/>
            <param name="initial_pose_a"            value="$(arg r4_yaw)" if="$(eval map == 'empty')"/>
            <remap from="/scan" to="$(arg robot4)/scan"/>
                
            <param name="global_frame_id" value="/map"/>
            <param name="odom_frame_id" value="$(arg robot4)/odom"/>
            <param name="base_frame_id" value="$(arg robot4)/base_footprint"/>    
        
        </node>

        <!--  ************** Navigation ***************  -->   
        <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
            <remap from="map" to="/map"/>
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="global_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/costmap_common_params.yaml" command="load" ns="local_costmap" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/local_costmap_params.yaml" command="load" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/sr7c_ls_param/global_costmap_params.yaml" command="load" />        
            
        <!-- Global Planner -->
            <param name="base_global_planner" value="global_planner/GlobalPlanner" />
            <param name="planner_frequency" value="1.0" />
            <param name="planner_patience" value="5.0" />
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/global_planner_params.yaml" command="load" />

            <param name="global_costmap/robot_base_frame" value="/$(arg robot4)/base_footprint"/>
            <param name="local_costmap/robot_base_frame" value="/$(arg robot4)/base_footprint"/>

            <!-- Local Planner -->
            <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS" />

            <!-- Our carlike robot is not able to rotate in place -->
            <param name="clearing_rotation_allowed" value="false" /> 

            <!-- external controller -->
            <remap from="cmd_vel" to="/fake_cmd" />    
        </node>
        
        <!--  ************** look a head **************  -->
        <node name="Look_a_head_Node" pkg="gazebo_rl_test" type="Look_a_head_Node" output="screen">
            <rosparam file="$(find gazebo_rl_test)/launch/Simulation/params/lookahead.yaml" command="load" />
            <param name="odom_topic"            value="/$(arg robot4)/odom"/>
            <param name="plan_topic"            value="/$(arg robot4)/move_base/GlobalPlanner/plan"/>
            <param name="goal_topic"            value="/$(arg robot4)/move_base_simple/goal"/>
            <param name="amcl_topic"            value="/$(arg robot4)/amcl_pose"/>
            <param name="lah_point_topic"       value="/$(arg robot4)/lookAhead_point"/>
            <param name="lah_marker_topic"      value="/$(arg robot4)/LookAhead/path_marker"/>
            <param name="odom_frame"            value="/$(arg robot4)/odom"/>

        </node>

    </group>


    <!--  ************** Visualisation **************  -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find gazebo_rl_test)/launch/Simulation/params/rviz_multi_4agent.rviz"/>
</launch>
